src/VisualStudio/Core/Def/Implementation/Utilities/ServiceInitializer.cs

```diff
@@ -0,0 +1,84 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.VisualStudio.Shell;
+using Roslyn.Utilities;
+using Task = System.Threading.Tasks.Task;
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation.Utilities
+{
+    internal class ServiceInitializer<TInterface, TService> where TInterface : class
+    {
+        private readonly SemaphoreSlim _gate = new SemaphoreSlim(initialCount: 1);
+
+        private readonly Shell.IAsyncServiceProvider _serviceProvider;
+        private readonly bool _uiThreadRequired;
+
+        private Func<TInterface, Task> _initializer;
+        private TInterface _service;
+
+        public ServiceInitializer(Shell.IAsyncServiceProvider serviceProvider, Func<TInterface, Task> initializer, bool uiThreadRequired)
+        {
```

Comment:
>{ [](start = 8, length = 1)

Consider checking that `serviceProvider` and `initializer` are not null and throw if they are. Better to find out here than later on.